runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java

```diff
@@ -201,56 +201,66 @@ public int hashCode() {
    */
   public class Reader extends BoundedSource.BoundedReader<T> {
     private long recordsRead = 0L;
-    private Instant endTime;
+    private Instant readEndTime;
     private final FluentBackoff backoffFactory;
-    private final UnboundedSource.UnboundedReader<T> reader;
+    private final UnboundedSource.UnboundedReader<T> unboundedReader;
     private boolean started;
 
-    private Reader(UnboundedSource.UnboundedReader<T> reader) {
-      this.reader = reader;
+    private Reader(final UnboundedSource.UnboundedReader<T> unboundedReader) {
+      this.unboundedReader = unboundedReader;
       backoffFactory =
           FluentBackoff.DEFAULT
               .withInitialBackoff(Duration.millis(10))
               .withMaxBackoff(maxReadTime.minus(1))
               .withMaxCumulativeBackoff(maxReadTime.minus(1));
     }
 
+    private boolean startIfNeeded() throws IOException {
+      return !started && ((started = true) && unboundedReader.start());
+    }
+
+    private void prepareForNewBatchReading() {
+      readEndTime = Instant.now().plus(maxReadTime);
+      recordsRead = 0L;
+    }
+
     @Override
     public boolean start() throws IOException {
-      LOG.debug("MicrobatchReader-{}: Starting a microbatch read from an unbounded source with a "
-          + "max read time of {} msec, and max number of records {}.", splitId, maxReadTime,
-              maxNumRecords);
-      endTime = Instant.now().plus(maxReadTime);
-      // Since reader is reused in microbatches only start it if it has not already been started.
-      if (!started) {
-        started = true;
-        if (reader.start()) {
-          recordsRead++;
-          return true;
-        }
-      }
-      return advanceWithBackoff();
+      LOG.debug(
+          "MicrobatchReader-{}: Starting a microbatch read from an unbounded source with a max "
+              + "read time of {} millis, and max number of records {}.",
+          splitId,
+          maxReadTime,
+          maxNumRecords);
+
+      prepareForNewBatchReading();
```

Comment:
@dhalperi , thanks for taking a look.

As of [BEAM-1294](https://issues.apache.org/jira/browse/BEAM-1294), readers are cached and thus maintain their state across multiple (Spark) batches. In order to have the `recordsCount` count elements per-batch, as opposed counting them per the reader's entire life-cycle, it needs to be re-set to `0`.

 At the moment the best place to perform this reset seems to be the `Reader#start()` method, however, as you (and the javadoc) point out, the `Reader#start()` method was originally designed to be called only.

I believe the motivation behind [BEAM-1294](https://issues.apache.org/jira/browse/BEAM-1294) was to avoid creating new readers for each new batch, as potentially there could be sources for which this would be a costly.

Are you aware of other sources / scenarios where readers are cached in a similar manner?